operators/pkg/custom-metrics-server/server.go

```diff
@@ -0,0 +1,83 @@
+// Copyright 2020-2022 Politecnico di Torino
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package custom_metrics_server contains the main logic and helpers
+// for the crownlabs custom-metrics-server component.
+package custom_metrics_server
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/go-logr/logr"
+	"google.golang.org/grpc"
+)
+
+// Server interface.
+type Server struct {
+	// MetricsScraperPeriod indicates the update interval for CustomMetrics update.
+	MetricsScraperPeriod time.Duration
+	grpcServer           *grpc.Server
+	Port                 int
+	Log                  logr.Logger
+	UnimplementedMetricsServerServer
+}
+
+// Init metricsScraper and RPC server.
+func (metricsServer Server) Init() error {
+	lis, err := net.Listen("tcp", fmt.Sprintf("%s%d", "0.0.0.0:", metricsServer.Port))
+	if err != nil {
+		return err
+	}
+
+	metricsServer.grpcServer = grpc.NewServer()
+	RegisterMetricsServerServer(metricsServer.grpcServer, &metricsServer)
+
+	log := metricsServer.Log.WithValues("listenerAddress", lis.Addr())
+	log.Info("Custom Metrics gRPC Server started")
+
+	go runMetricsScraper(log.WithName("metrics-scraper"), metricsServer.MetricsScraperPeriod)
```

Comment:
The idea is to start scraping metrics right before the MetricsServer starts serving clients. This sequence of goroutine calls avoids situation where empty resoponses are given to clients just because no metrics are collected by the server.

Reply:
Go routines are like threads, you are not guaranteed they are executed in order.